Switch jsxstyle and jsxstyle-utils to TypeScript

meyer opened the pull request:
This is a first pass at switching jsxstyle to TypeScript. Main motivation is that maintaining separate type definition files seems like it would lead to inconsistencies.

@appsforartists and @steadicat, I’d love some input if you’ve got the time.

steadicat commented on packages/jsxstyle-utils/src/componentStyles.ts:
```diff
@@ -1,3 +1,7 @@
+export interface StyleObject {
+  [key: string]: { [key: string]: string | number } | null;
```
For our project, I lifted [CSS types](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L817) from `@types/react`. It's pretty important for `jsxstyle` to have strict types for CSS props. Otherwise any prop on any component goes, which would hide a ton of errors.

steadicat commented on packages/jsxstyle-utils/src/getStyleCache.ts:
```diff
@@ -2,6 +2,26 @@ import getStyleKeysForProps from './getStyleKeysForProps';
+}
+
+export interface StyleCache {
+  reset: { (): void };
```
Is this the same as `reset(): void`?

meyer replied:
@steadicat heh, yeah. can you tell i’m new to TypeScript? 😅

jaredpalmer commented:
Awwww yeah

steadicat commented on packages/jsxstyle/jsxstyle.tsx:
```diff
@@ -1,21 +1,30 @@
+  const tagName = 'div';
+
+  return class JsxstyleComponent extends React.Component<
+    JsxstyleProps & { [key: string]: any }
```
I highly recommend removing `{[key: string]: any}` here, for the reason mentioned above.

steadicat commented on packages/jsxstyle/jsxstyle.tsx:
```diff
@@ -1,21 +1,30 @@
-  }
+export interface JsxstyleProps {
+  className?: string;
+  component?: string | React.ComponentClass | React.SFC;
```
Use `keyof JSX.IntrinsicElements` instead of `string`.

meyer replied:
🙏 thank you

steadicat commented on packages/jsxstyle/jsxstyle.tsx:
```diff
@@ -1,21 +1,30 @@
+  className?: string;
+  component?: string | React.ComponentClass | React.SFC;
+  mediaQueries?: { [key: string]: string };
+  props?: { [key: string]: any };
```
Use `React.AllHTMLAttributes<T> & React.ClassAttributes<T>`.

steadicat commented on packages/jsxstyle/jsxstyle.tsx:
```diff
@@ -1,21 +1,30 @@
+export interface JsxstyleProps {
+  className?: string;
+  component?: string | React.ComponentClass | React.SFC;
+  mediaQueries?: { [key: string]: string };
```
Enumerate all possible media queries here if possible.

meyer replied:
The main problem here is that the media query key can be anything the user specifies. The API does what it needs to do but I’m not super happy with how open-ended it makes things.

steadicat commented on packages/jsxstyle-utils/src/componentStyles.ts:
```diff
@@ -6,4 +10,4 @@ export default {
   Col: { display: 'flex', flexDirection: 'column' },
   Grid: { display: 'grid' },
-};
+} as StyleObject;
```
Why is this needed? TS should be able to infer a better type for this.

meyer replied:
@steadicat good call, I’ll remove it.

steadicat commented on packages/jsxstyle-utils/src/addStyleToHead.ts:
```diff
@@ -23,18 +23,19 @@ if (canUseDOM && !styleElement) {
-  if (canUseDOM) {
+export default function addStyleToHead(rule: string) {
+  if (styleElement) {
+    const sheet = styleElement.sheet as CSSStyleSheet;
```
Why is the cast needed here?

meyer replied:
@steadicat the cast feels gross but without it I get `Property 'insertRule' does not exist on type 'StyleSheet'`. Apparently (according to https://github.com/Microsoft/TypeScript/issues/15319) this issue occurs because `styleElement.sheet` might be one of several types.

steadicat replied:
That’s weird, the [type definition of LinkStyle.sheet](https://github.com/Microsoft/TypeScript/blob/master/lib/lib.dom.d.ts#L13685) only has one type.

meyer replied:
Yeah, it’s funky. There’s currently only one type of `StyleSheet` in the spec (`CSSStyleSheet`) but there might be more than one in the future.

https://github.com/Microsoft/TypeScript/issues/15319#issuecomment-296584050
https://drafts.csswg.org/cssom/#the-stylesheet-interface

steadicat commented:
BTW `React.CSSProperties` has a string index accessor, which sucks. I forked it and removed it. I would also replace all the `| any` with actual allowed values.

meyer commented:
@steadicat I’m looking into a way of auto-generating `CSSProperties` with jsxstyle-specific additions, like the pseudoclass and pseudoelement prefixes. If you’ve got ideas let me know! For now I’m probably going to parse the relevant parts of `@types/react` with babylon and add the prefixed style props at the AST level.

The main issue (as I mentioned above) is that the media query prefix is user-specified and can therefore be _anything_. I’m not sure what to do about that type-wise other than to either use an index accessor (😩) or rethink the media query feature entirely.

meyer commented on packages/jsxstyle-utils/src/getStyleCache.ts:
```diff
@@ -14,17 +34,17 @@ function alreadyInjected() {
 }
 
-function getStringHash(key) {
+function getStringHash(key: string, _: {}) {
```
Is there an official way to ignore a second parameter? It’s passed to `getClassNameForKey` on line 72 and I get why TS would complain about function signature being different, but an unused underscore param seems hella nasty. It gets minified out so I’m not too stressed about it, but just seems like there’s got to be a better way to handle this.

jaredpalmer replied:
Yeah prefix name with underscore and it will be ignored if unused

steadicat replied:
You can use function overloads for this:
```ts
function getStringHash(key: string, props: any);
function getStringHash(key: string) {
  ...
}
```

meyer commented:
I added a script a little while ago that extracts out `CSSProperties` from `@types/react`, and I just added prefixing to the script locally. 59 prefixes are added to each prop in the `CSSProperties` interface. The generated file is 23K lines—a few bytes short of a megabyte (!!!).

![image](https://user-images.githubusercontent.com/13781/33006848-c891353e-cd80-11e7-9565-fcb3eccef0d8.png)

appsforartists commented on packages/jsxstyle-utils/src/getStyleCache.ts:
```diff
@@ -2,6 +2,20 @@ import getStyleKeysForProps from './getStyleKeysForProps';
 import stringHash from './stringHash';
 
+export interface InsertRuleCallback {
+  (rule: string, props?: {}): boolean | void;
```
I feel like you could get fancy here and type `props` more tightly, the same way that `document.createElement` returns a different type based on which string you pass it.  I think it's something like:


```typescript
export interface InsertRuleCallback<K extends keyof RuleToProps> {
  (rule: K, props?: RuleToProps[K]): boolean | void;
}
```

and it looks like this is just a function type.  I'd have expected

```ts
export type InsertRuleCallback<K extends keyof RuleToProps> = (rule: K, props?: RuleToProps[K]) => boolean | void;
```
but that's just a style thing.  I've never seen `interface` used to represent a single function signature before.

steadicat replied:
Agree that this should probably be:
```ts
function InsertRuleCallback(rule: string, props?: {}): boolean | void;
```

meyer replied:
@steadicat does this look right?

```typescript
type InsertRuleCallback = (rule: string, props?: {}) => boolean | void;
```

@appsforartists for now, in order to get this PR landed, I’m going to leave this as the not-so-great generic props type, but I’d love a PR to improve this if you have the time. I’m reaching the limits of what I can google regarding TypeScript 😅

appsforartists commented on packages/jsxstyle-utils/src/getStyleCache.ts:
```diff
@@ -2,6 +2,20 @@ import getStyleKeysForProps from './getStyleKeysForProps';
+
+export interface StyleCache {
+  reset(): void;
+  getClassName(props: { [key: string]: any }, classNameProp?: string): string;
```
In my codebases, I like to have a `Dict` type:

```ts
export type Dict<T> = {
  [key: string]: T,
};
```
which makes this a little easier to express:

```ts
  getClassName(props: Dict<any>, classNameProp?: string): string;
```

meyer replied:
ooh, this is neat.

appsforartists commented on packages/jsxstyle-utils/src/getStyleCache.ts:
```diff
@@ -14,17 +28,17 @@ function alreadyInjected() {
 }
 
-function getStringHash(key) {
+function getStringHash(key: string, props: {}) {
```
You're not using `props` here.  I take it this is because you have call sites that pass two params.

The 'correct' way to express this is probably something like:

```ts
export type GetClassNameForKey = (key: string, props: {}) => string;

let getClassNameForKey = getStringHash as GetClassNameForKey;
```

but if you wanted something quicker and more concise, you could make `props` optional:

```typescript
function getStringHash(key: string, props?: {}): string {
```

meyer replied:
Yeah, I didn’t use an optional param because it didn’t seem to change anything (probably because I haven’t hooked up tslint yet). @steadicat mentioned function overloads earlier (https://github.com/smyte/jsxstyle/pull/89#discussion_r151855318). All these options seem kinda gross, but function overloads seem the least gross right now.

appsforartists replied:
Personally, I think having a `GetClassNameForKey` type is the most semantic and least gross.  It's a bit more boilerplate, but it expresses your intent clearly (you have a callback that passes both arguments, you want to set it to one that only needs one argument, and you use casting to tell TS that the arity mismatch is OK.)

appsforartists commented on packages/jsxstyle-utils/src/getStyleCache.ts:
```diff
@@ -14,17 +28,17 @@ function alreadyInjected() {
   let pretty = false;
 
-  const styleCache = {};
+  const styleCache: StyleCache = <any>{};
```
`({} as any)` is easier to read.  I don't often see `<>` used for casting.

But more holistically, I don't think you need to cast this at all - it's just sidestepping the type checker.  Can you just refactor it to

```typescript
const styleCache: StyleCache = {
  reset() {
    _classNameCache = {}
  }
  // ...
}
```

meyer replied:
Yeah, refactoring is the way to go I think. I was trying to refactor as little as possible to get a first version out ASAP, but in this case it’s quick and easy.

appsforartists commented on packages/jsxstyle-utils/src/getStyleKeysForProps.ts:
```diff
@@ -35,7 +35,7 @@ const specialCaseProps = {
 };
 
-export default function getStyleKeysForProps(props, pretty = false) {
+export default function getStyleKeysForProps(props: any, pretty = false) {
```
I wonder again if there is a tighter way to type `props`.

appsforartists commented on packages/jsxstyle-utils/src/getStyleKeysForProps.ts:
```diff
@@ -52,7 +52,14 @@ export default function getStyleKeysForProps(props, pretty = false) {
 
   // return value
-  const styleKeyObj = {};
+  const styleKeyObj: {
```
You should extract this type and explicitly use it as the return value on the function.  That will make the function signature easier to skim, and keep you from having to cast an empty dictionary to match a type it doesn't yet implement.

meyer replied:
What type would I assign to `styleKeyObj` in that case? I’m not sure how to initialise an empty object with a type in TypeScript without making all the keys optional (which seems wrong to me? idk tho).

appsforartists replied:
I'm not sure if you would need to set it to anything.  I think the type could be inferred in the function body and explicitly set at the function boundary.

```typescript
export type StyleKey = {
  styles: string,
  mediaQuery?: string,
  pseudoclass?: string,
  pseudoelement?: string,
};

export default function getStyleKeysForProps(props: any, pretty = false): Dict<StyleKey> {
  const result = {};
  // build result
  return result
}
```

meyer replied:
I was getting this error when I tried that earlier:
```
Type '{}' is not assignable to type 'StyleKeyObject | null'.
```

appsforartists replied:
Maybe I was too optimistic about what it could infer.  I was sleepy and on a plane when I wrote that (and I'm still sleepy and just-off a plane), so I hadn't actually tested it.

In that case, probably:

```typescript
export type StyleKey = {
  styles: string,
  mediaQuery?: string,
  pseudoclass?: string,
  pseudoelement?: string,
};

export default function getStyleKeysForProps(props: any, pretty = false): Dict<StyleKey> {
  const result: Dict<StyleKey> = {};
  // build result
  return result
}
```
You don't even have to `Partial<>` it, because a `Dict` can have no members and still be a `Dict`.

On a related note, I think you're always supposed to give explicit types to the arguments and return value of any exported function.  I know that's true in Flow, but I only presume so in TS.  It's supposed to simplify the amount of work the typechecker has to do if you annotate at the boundaries, which makes type-checking easier to parallelize.

meyer replied:
Heh, no problem. Yeah, that makes sense to use an explicit return type. I wasn’t wild about the inline multiline type definition anyway. Much nicer to have it top-level.

appsforartists commented on packages/jsxstyle/jsxstyle.tsx:
```diff
@@ -1,21 +1,41 @@
+
+function factory(
+  displayName: string,
+  defaultProps?: { [key: string]: React.ReactText } | null
```
This is why `Dict` is cool. 😄

appsforartists commented on rollup.config.js:
```diff
@@ -11,50 +11,25 @@ const external = [
-  plugins: [
-    require.resolve('babel-plugin-transform-class-properties'),
-    require.resolve('babel-plugin-transform-object-assign'),
-    require.resolve('babel-plugin-external-helpers'),
```
Babel is awesome, but it's nice to not have to worry about which not-yet-standardized features you're opting into.

meyer replied:
Yeah, in this case, Babel wasn’t doing anything, since TypeScript handles JSX, `Object.assign` (at least with JSX spreads), and ES6 classes.

steadicat commented on packages/jsxstyle/jsxstyle.tsx:
```diff
@@ -1,21 +1,41 @@
+  className?: string;
+  component?: ComponentProp;
+  mediaQueries?: { [key: string]: string };
+  props?: { [key: string]: any };
```
`props` on the public interface really shouldn't be a `{[key: string]: any}`. How about something like this:
```ts
export type AnyComponent<ComponentProps> =
  | keyof JSX.IntrinsicElements
  | React.ComponentClass<ComponentProps>
  | React.SFC<ComponentProps>;

export type JsxstyleProps<ComponentProps> = {
  className?: string;
  component?: AnyComponent<ComponentProps>;
  mediaQueries?: {[key: string]: string};
  props?: ComponentProps;
  style?: CSSProperties;
} & CSSProperties;
```

meyer replied:
@steadicat I can dig it. The only thing I’m not sure about is what I pass to `JsxstyleProps` when I specify it as a type. For example:
```typescript
constructor(props: JsxstyleProps<???>) { // <--- what goes there?
  super(props);
  this.component = props.component || tagName;
  this.className = cache.getClassName(props, props.className);
}```

Sorry, this is probably basic stuff. Still learning this TypeScript biz.

steadicat replied:
The generic param to `JsxstyleProps` should bubble all the way up to the caller. So, e.g., you’d define `Block` like this:

```ts
class Block<InnerComponentProps> extends React.Component<JsxstyleProps<InnerComponentProps>> {
  constructor(props: JsxstyleProps<InnerComponentProps>) {
    ...
  }
  ...
}
```

This way each call site determines which props are allowed. Ideally `InnerComponentProps` will be inferred by the value of the `component` prop. Should happen magically with the type definition I gave you above. Makes sense?

meyer replied:
Ahhhh yep that makes sense. This is cool!

meyer replied:
@steadicat does it make sense for `AnyComponent`’s props to extend `ComponentPropProps`? I’d like TypeScript to complain if a component specified as the `component` prop doesn’t accept the `className` prop (and perhaps the `style` prop as well).

Example:

```typescript
export interface ComponentPropProps {
  className?: any;
  style?: CSSProperties;
  [key: string]: any;
}

export type AnyComponent<Props extends ComponentPropProps> =
  | keyof JSX.IntrinsicElements
  | React.ComponentClass<Props>
  | React.SFC<Props>;

// ...
```

I think this does what I expect, but I might be wrong. React might already set `style` to `React.CSSProperties` and `className` to something like `any`? I’m not 100% sure.

steadicat replied:
Yes! You got this! The index accessor on `ComponentPropProps` shouldn't be needed. Just specify the props you need the component to accept. I believe `className` is something like `string | null | undefined | false`. VSCode is great when you're discovering this stuff. You can hover over anything to see the inferred type, and jump to definition to see the full type declaration. Also, nit: the type of `style?: CSSProperties` should probably be the native React one, not the Jsxstyle kind.

steadicat commented on packages/jsxstyle/preact/jsxstyle-preact.tsx:
```diff
@@ -0,0 +1,67 @@
+
+export interface JsxstyleProps extends CSSProperties {
+  className?: string;
+  component?: keyof JSX.IntrinsicElements | React.ComponentClass | React.SFC;
```
Does this work? Using `React` types in `preact`?

meyer replied:
Nope this is an error. I pasted a bit too much.

steadicat commented on packages/jsxstyle/preact/jsxstyle-preact.tsx:
```diff
@@ -0,0 +1,67 @@
+  mediaQueries?: { [key: string]: string };
+  props?: { [key: string]: any };
+  style?: string;
+  children: (JSX.Element | JSX.Element[] | string)[];
```
Can't children also be `null` or `undefined` or `false` or an array of any of the above? Does `preact` not have a `Children` type you can use?

meyer replied:
I haven’t spent more than 30 minutes on the preact version so I’m not 10)% sure. I definitely hope preact has a Children type. I pulled this from a `children` key in preact’s type definition file.

meyer commented:
I gave this PR a bit of thought on my bike ride home. I’d prefer to land it in an imperfect but permissive state—i.e., not breaking everyone‘s builds—than to try to perfect TypeScript support in one PR. Future commits and PRs can make the generic type definitions more specific and fix my noob TypeScript code.

If you all think that’s a bad idea, do let me know. Otherwise I’ll aim to have this landed tomorrow or Wednesday, once I’ve got `jsxstyle/preact` in good shape.

meyer commented:
Also, that way you all can get commit credit for your contributions! I feel bad copying and pasting your code into my editor 😁

meyer commented:
Preact version is a bit gross but it builds. It’s shippable (imo) but I’d love for someone who uses Preact to improve it at some point in the future.

I think this is ready to land. Lay your final feedback upon me.

appsforartists commented on packages/jsxstyle-utils/src/getStyleKeysForProps.ts:
```diff
@@ -35,7 +35,19 @@ const specialCaseProps = {
 };
 
-export default function getStyleKeysForProps(props: any, pretty = false) {
+export type StyleKeyObj = {
```
This is a `Dict`!

meyer replied:
🙏 thx

appsforartists commented on packages/jsxstyle-utils/src/getStyleKeysForProps.ts:
```diff
@@ -52,14 +64,7 @@ export default function getStyleKeysForProps(props: any, pretty = false) {
-      pseudoelement?: string;
-    };
-  } & { classNameKey: string } = <any>{};
+  const styleKeyObj: StyleKeyObj = <any>{};
```
Can you cast this to `StyleKeyObj` instead of `any`?

(I still think you should be using `as` for casting, but that's a style thing.)

meyer replied:
Yep, that works. Thanks!

appsforartists replied:
I might even call it `StyleKeyDict` to be consistent with your naming.

appsforartists commented on packages/jsxstyle-utils/src/getStyleCache.ts:
```diff
@@ -16,7 +17,8 @@ function alreadyInjected() {
+function getStringHash(key: string) {
   return '_' + stringHash(key).toString(36);
 }
 
```
`getStyleCache` is missing a return type annotation

meyer replied:
Yeah, it’s inferred in the second case.

meyer replied:
Thought about this a bit more. It’s probably best to be explicit.

appsforartists commented:
:tada:!

meyer commented:
Published as `jsxstyle@2.0.5` and `jsxstyle-utils@2.0.3`.
